### PR TITLE
Changed path_item_finder to detect more complex path template patterns

### PR DESCRIPTION
**Pull request by makdad**

### CODE REVIEW

**Comment by ota42y** [comment]
(from https://github.com/interagent/committee/issues/261)

Hi! 
I think there is no problem with the method interface, style, etc... 👍 

But there is many bugs
And I think we need to make a big change to fix this problem...
(It seems like a more difficult problem than we expected)
```
f = OpenAPIParser::PathItemFinder.new([])
f.parse_path_parameters('{p1}.json', 'foo-.jso')
# Should be return nil but {"p1"=>"foo-"}

f.parse_path_parameters('{p1}schedule', 'adminsschedule')
# Should be return 'admins' but {"p1"=>"admin"}

f.parse_path_parameters('{p1}schedule', 'usersschedule')
# Should be return 'usres' but {"p1"=>"u"}

f.parse_path_parameters('{p1}-{p2}.json', 'foo.json')
# Should be return nil but got errors
```

**Comment by makdad** [comment]
@ota42y thanks a lot, I have added a commit with failing specs to cover your feedback.  Will work on the implementation for now.  I'll close this PR until we have it covered.

**Comment by ota42y** [comment]
I think we can convert path parameter to regexp 🤔 
( `{p1}-{p2}.json` -> `(.+)-(.+)\.json` )

**Comment by makdad** [comment]
@ota42y your hunch was correct, this code is so much simpler by relying on the underlying Ruby regex engine. Thank you for the idea!  I'm ready for review again; if you think this is reasonable, I can open a PR to the main repository.

**Comment by makdad** [comment]
@ota42y The only thing I am not certain about is if we need to `Regex.escape` the non-parameter portions of the path. I did some various tests via unit testing, but it all seemed to work as-expected.  I think the more "invalid" combinations that might generate the situation would not be valid URLs, so it is probably OK?

File: lib/openapi_parser/path_item_finder.rb [hunk 1]
```diff
@@ -38,7 +38,33 @@ def initialize(path_item_object, operation_object, original_path, path_params)
     end
   end
 
+  def parse_path_parameters(schema_path, request_path)
+    parameters = path_parameters(schema_path)
+    return nil if parameters.empty?
+
+    regex = schema_path.dup
+    parameters.each do |parameter|
+      regex = regex.gsub(parameter, "(?<#{param_name(parameter)}>.+)")
```
Comment (ota42y) [inline lib/openapi_parser/path_item_finder.rb]: 👍 👍 👍

**Review by ota42y — CHANGES_REQUESTED** [review]
I think there is very good code.

But your concerns need to be checked.
Please check these test case.

- `{id}.json` and `123-json`
  - dose not match.
- `{id>}.json` and `1.json`
  - Parameter name is "`id>`" and that value is `1`
  - The OpenAPI dosen't define valid character so I think any characters without `{}` is allowed 
    - ( Please tell me if I misread.. I'm not good at English)
  -  https://github.com/OAI/OpenAPI-Specification/blob/3.0.3/versions/3.0.3.md#path-templating

**Comment by makdad** [comment]
@ota42y Thanks for pointing this out; I've updated the code to escape the non-parameter components.  Also, don't worry about your English - it's great!  In the worst case, we can also speak Japanese 👍

**Comment by ota42y** [comment]
Sorry for my late reply, I think this PR is perfect !!! 🎉 🎉 🎉 
Thank you for your kind words 👍

**Pull request merged**